include/zio/zio_device.h

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright (c) 2019 Thomas Burdick <thomas.burdick@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_ZIO_DEVICE_H_
+#define ZEPHYR_INCLUDE_ZIO_DEVICE_H_
+
+#include <zephyr/types.h>
+#include <device.h>
+
+/**
+ * A ZIO device provides attributes, channels, and triggers for a device.
+ *
+ * Channels are data streams of input or output with attributes themselves which may be
+ * subscribed to.
+ */
+struct zio_channel_desc
+{
+  const char *name;
+  const u16_t kind;
+  const enum direction direction;
+  const u8_t bitwidth;
+  const u8_t bitshift;
+  const enum container container;
+  const enum byteorder byteorder;
+  const u8_t byteshift;
+};
+
+
+/**
+ * @brief A type tagged variant for attributes
+ */
+struct zio_variant
+{
+  enum {
+        float_,
+        double_,
+        u8_,
+        u16_,
+        u32_,
+        u64_,
+        s8_,
+        s16_,
+        s32_,
+        s64_,
+  } tag;
+  union {
+    float float_val;
+    double double_val;
+    u8_t  u8_val;
+    u16_t u16_val;
+    u32_t u32_val;
+    u64_t u64_val;
+    s8_t  s8_val;
+    s16_t s16_val;
+    s32_t s32_val;
+    s64_t s64_val;
+  } value;
+}
+
+/**
+ * @brief Function to set a channel attribute for a device
+ *
+ * @return -EINVAL if an invalid attribute id, type, or range was given, 0 otherwise
+ */
+typedef int (*zio_channel_set_attr_t)(struct device *dev, const struct zio_channel_desc *channel, const uint32_t id, const struct zio_variant attr);
+
+/**
+ * @brief Function to get a channel attribute for a device
+ *
+ * @return -EINVAL if an invalid attribute id was given, 0 otherwise
+ */
+typedef int (*zio_channel_get_attr_t)(struct device *dev, struct zio_channel_desc *channel, uint32_t id, struct zio_variant *attr);
+
```

Comment:
What mechanism do you envision to be able to iterate through the list of attributes (at the app level) when multiple instances are present?

Reply:
Concerning traversing the attribute (etc.) collection, would having the list of attributes as a singly-linked list be inappropriate here?

`SLIST` (see [queue.h](https://github.com/apache/mynewt-core/blob/master/sys/sys/include/sys/queue.h)) is used quite heavily in Mynewt, for example, and it does make life easier in some situations and enable dynamic addition or removal of attributes at run-time. For example, see: https://github.com/apache/mynewt-core/blob/master/hw/sensor/include/sensor/sensor.h#L278

Reply:
I was thinking about that further, I think like channels it might be worth creating a static array per device type that describes each of the attributes. You could iterate through the each attribute that way then.

Reply:
One advantage of singly-linked lists is that you can dynamically add items, but a static list is also easy to implement in code and understand. I don't know if there is a super compelling case for keeping the list dynamic, though, so I don't really have strong feelings either way myself.

Reply:
I haven't personally come across an instance where I'd need a dynamic number of attributes for this sort of scenario, I'd like to go forward with the static arrays for attributes for now

Reply:
Agreed ... better to start simple and complicate things later on a genuinely as-needed basis. I can think of situations where we want to dynamically disable/enable channels (for power reasons, etc.), but that's just a field so no changes to the list itself.